Code/GraphMol/CIPLabeler/rules/Rule4b.cpp

```diff
@@ -0,0 +1,301 @@
+//
+//
+//  Copyright (C) 2020 Schrödinger, LLC
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+
+#include <algorithm>
+#include <list>
+
+#include <RDGeneral/Invariant.h>
+
+#include "Rule4b.h"
+
+#include "../Digraph.h"
+#include "Pairlist.h"
+
+namespace RDKit {
+namespace CIPLabeler {
+
+Rule4b::Rule4b() = default;
+
+Rule4b::Rule4b(Descriptor ref) : d_ref{ref} {}
+
+std::vector<Descriptor>
+Rule4b::getReferenceDescriptors(const Node *node) const {
+  std::vector<Descriptor> result;
+  auto prev = initialLevel(node);
+  while (!prev.empty()) {
+    for (const auto &nodes : prev) {
+      if (getReference(nodes, result)) {
+        return result;
+      }
+    }
+    prev = getNextLevel(prev);
+  }
+  return {};
+}
+
+int Rule4b::compare(const Edge *a, const Edge *b) const {
+  const auto &aBeg = a->getBeg();
+  const auto &aEnd = a->getEnd();
+  const auto &bBeg = b->getBeg();
+  const auto &bEnd = b->getEnd();
+  if (aBeg->getDigraph()->getCurrentRoot() != aBeg ||
+      bBeg->getDigraph()->getCurrentRoot() != bBeg) {
+    if (d_ref == Descriptor::NONE) {
+      return 0;
+    }
+    Descriptor aDesc = aEnd->getAux();
+    Descriptor bDesc = bEnd->getAux();
+    if (aDesc != Descriptor::NONE && bDesc != Descriptor::NONE &&
+        aDesc != Descriptor::ns && bDesc != Descriptor::ns) {
+      bool alike = PairList::ref(d_ref) == PairList::ref(aDesc);
+      bool blike = PairList::ref(d_ref) == PairList::ref(bDesc);
+      if (alike && !blike) {
+        return +1;
+      }
+      if (blike && !alike) {
+        return -1;
+      }
+    }
+    return 0;
+  } else {
+    auto list1 = newPairLists(getReferenceDescriptors(aEnd));
+
+    auto list2 = newPairLists(getReferenceDescriptors(bEnd));
+
+    if (list1.empty() != list2.empty()) {
+      throw std::runtime_error(
+          "Substituents should be topologically equivalent!");
+    }
+    if (list1.size() == 1) {
+      return comparePairs(aEnd, bEnd, list1[0].getRefDescriptor(),
+                          list2[0].getRefDescriptor());
+    } else if (list1.size() > 1) {
+      for (auto &plist : list1) {
+        fillPairs(aEnd, plist);
+      }
+      for (auto &plist : list2) {
+        fillPairs(bEnd, plist);
+      }
+
+      std::sort(list1.rbegin(), list1.rend());
+      std::sort(list2.rbegin(), list2.rend());
+
+      for (auto i = 0u; i < list1.size(); ++i) {
+        int cmp = list1[0].compareTo(list2[0]);
```

Comment:
shouldn't this be:
```suggestion
        int cmp = list1[i].compareTo(list2[i]);
```

Reply:
That totally makes sense. I fixed it, and I don't see any changes in the validation, though.

And this is also an upstream bug: https://github.com/SiMolecule/centres/blob/33e09a3a40bd766008acc18fdd939209b529b82c/core/src/main/java/com/simolecule/centres/rules/Rule4b.java#L384

Reply:
reported